### PR TITLE
SmartData editors.

### PR DESCRIPTION
**Pull request by pelekon**

### CODE REVIEW

File: WoWDatabaseEditor.Common/WDE.Common/Providers/IDataDefinitionsProvider.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+using WDE.Module.Attributes;
+using WDE.Common.Managers;
+
+namespace WDE.Common.Providers
+{
+    [NonUniqueProvider]
+    public interface IDataDefinitionsProvider
+    {
+        string GetDataCategoryName();
+        IEnumerable<IDataDefinitionPresenter> GetDataDefinitionPresenters();
```
Comment (BAndysc) [inline WoWDatabaseEditor.Common/WDE.Common/Providers/IDataDefinitionsProvider.cs:11]: Can we change name? Presenter sound too much like „view”. GetDataDefinitionEditors?

File: WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataTargetsEditorView.xaml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                Width="25" Height="25" VerticalAlignment="Center" HorizontalAlignment="Left" Margin="10, 0, 0, 10">
+            <Path HorizontalAlignment="Center" VerticalAlignment="Center" Stroke="{DynamicResource {x:Static SystemColors.ControlTextBrushKey}}"
+                  Data="M3.875,0 L5.125,0 5.125,3.875 9,3.875 9,5.125 5.125,5.125 5.125,9 3.875,9 3.875,5.125 0,5.125 0,3.875 3.875,3.875 3.875,0 z" />
+        </Button>
+        <ListView MinHeight="100" ItemsSource="{Binding Source.Parameters}" SelectedIndex="{Binding SelectedParamIndex}">
+            <ListView.ItemTemplate>
+                <DataTemplate>
+                    <Grid>
+                        <Grid.ColumnDefinitions>
+                            <ColumnDefinition />
```
Comment (BAndysc) [inline WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataTargetsEditorView.xaml:101]: Could you set MinWidth here? Something around 100 units maybe? To make it looks more like a table

File: WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataParameterEditorView.xaml [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                            
+                        </Style>
+                    </ListView.ItemContainerStyle>
+                </ListView>
+            </ScrollViewer>
+        </StackPanel>
+        <!--#endregion-->
+        <!--#region Bottom Bar-->
+        <Button Name="saveButton" Content="Save" Command="{Binding SaveValue}" HorizontalAlignment="Center" Width="50" Height="30" Margin="0,10, 0, 0">
+            <!--<Button.Style>
```
Comment (BAndysc) [inline WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataParameterEditorView.xaml:129]: Can you remove it if it is unused?

File: WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataEventsEditorView.xaml [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                            <ListView.ItemTemplate>
+                                <DataTemplate>
+                                    <Grid>
+                                        <Grid.ColumnDefinitions>
+                                            <ColumnDefinition />
+                                            <ColumnDefinition />
+                                            <ColumnDefinition />
+                                        </Grid.ColumnDefinitions>
+                                        <!-- Change that!! -->
+                                        <Label Grid.Column="0" Content="{Binding Description}" />
```
Comment (BAndysc) [inline WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataEventsEditorView.xaml:185]: Change to what? 👯

File: WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataParameterEditorViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            Source = new SmartDataParameterEditorData(in source);
+            AreValuesVisible = false;
+            AssignCurrentParamKey();
+            AddValue = new DelegateCommand(AddValueToParam);
+            DeleteValue = new DelegateCommand(DeleteValueFromParam);
+            SaveValue = new DelegateCommand(() => CloseOk?.Invoke());
+        }
+
+        public SmartDataParameterEditorData Source { get; }
+        public bool AreValuesVisible { get; private set; }
```
Comment (BAndysc) [inline WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataParameterEditorViewModel.cs:45]: Now, if this value change (for instance when changing parameter type), the UI doesn’t update. If you implement INotifyPropertyChanged for this type, will it update?

File: WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataDefinesListViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        public int SelectedItemIndex { get; set; }
+        public DelegateCommand<object> OnItemSelected { get; }
+        public DelegateCommand CreateNew { get; }
+        public DelegateCommand DeleteItem { get; }
+
+        public ObservableCollection<SmartGenericJsonData> DefinesItems { get; }
+
+        private void ShowEditorWindow(object action)
+        {
+            if (action is SmartGenericJsonData item)
```
Comment (BAndysc) [inline WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataDefinesListViewModel.cs:71]: object ;(

**Review by BAndysc — CHANGES_REQUESTED** [review]

File: WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataGroupsEditorViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+                var result = data.Name;
+                OpenNameEditingWindow(data.Name, out result);
+                if (!string.IsNullOrEmpty(result))
+                    data.UpdateName(result);
+            }
+        }
+
+        private void OpenNameEditingWindow(string source, out string name)
+        {
+            var v = new SmartDataGroupsInputView();
```
[the remark this answers is not in the document]
Comment (pelekon) [reply WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataGroupsEditorViewModel.cs:100]: Done

File: WoWDatabaseEditor.sln [hunk 8]
```diff
@@ -3,7 +3,7 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio Version 16
 VisualStudioVersion = 16.0.29728.190
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WoWDatabaseEditorCore", "WoWDatabaseEditor\WoWDatabaseEditorCore.csproj", "{99C9E82C-C594-446D-AA59-8FFBC43AD226}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WoWDatabaseEditorCore", "WoWDatabaseEditor\WoWDatabaseEditorCore.csproj", "{99C9E82C-C594-446D-AA59-8FFBC43AD226}"
```
[the remark this answers is not in the document]
Comment (pelekon) [reply WoWDatabaseEditor.sln:6]: No idea

File: WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataEventsEditorView.xaml (hunk 4 above)
Comment (pelekon) [reply WDE.SmartScriptEditor/Editor/Views/SmartDataEditors/SmartDataEventsEditorView.xaml:185]: Forgotten reminder xd

File: WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataParameterEditorViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    class SmartDataParameterEditorData
+    {
+        public string Name { get; set; }
+        public string Description { get; set; }
+        public string Type { get; set; }
+        public bool Required { get; set; }
+        public int DefaultVal { get; set; }
+        public ObservableCollection<BindableTuple<int, SelectOption>> Values { get; set; }
```
[the remark this answers is not in the document]
Comment (pelekon) [reply WDE.SmartScriptEditor/Editor/ViewModels/SmartDataEditors/SmartDataParameterEditorViewModel.cs:116]: Cuz KeyPairValue has key as readonly and I needed both values to be editable.

File: WDE.SmartScriptEditor/Providers/SmartDataPresentersProvider/SmartDataEditorsProvider.cs [hunk 10]
```diff
@@ -21,24 +21,27 @@ public enum SmartDataSourceMode
...
+                new SmartDataPresenter<SmartDataGroupsEditorViewModel>("Smart Data - Action Groups", new object[] { smartDataProvider, taskRunner, 
+                    messageBoxService, windowManager, SmartDataSourceMode.SD_SOURCE_ACTIONS }),
+                new SmartDataPresenter<SmartDataGroupsEditorViewModel>("Smart Data - Target Groups", new object[] { smartDataProvider, taskRunner, 
+                    messageBoxService, windowManager, SmartDataSourceMode.SD_SOURCE_TARGETS }),
             }.AsEnumerable();
         }
         public string GetDataCategoryName() => "Smart Data";
 
-        public IEnumerable<IDataDefinitionPresenter> GetDataDefinitionPresenters() => presenters;
+        public IEnumerable<IDataDefinitionPresenter> GetDataDefinitionEditors() => editors;
```
Comment (BAndysc) [inline WDE.SmartScriptEditor/Providers/SmartDataPresentersProvider/SmartDataEditorsProvider.cs:45]: How about changing IDataDefinitionPresenter name to editor as well ?

File: WDE.Parameters/Views/ParametersView.xaml [hunk 11]
```diff
@@ -3,6 +3,11 @@
              xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
              xmlns:prism="http://prismlibrary.com/"
              prism:ViewModelLocator.AutoWireViewModel="False">
+    <UserControl.Resources>
```
Comment (BAndysc) [inline WDE.Parameters/Views/ParametersView.xaml:6]: Is it added to fix color on List items?

**Pull request merged**